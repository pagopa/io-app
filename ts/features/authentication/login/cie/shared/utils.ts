import { Platform } from "react-native";
import { isDevEnv } from "../../../../../utils/environment";
import { SpidLevel } from "../utils";

export type WebViewLoginNavigationProps = {
  spidLevel: SpidLevel;
  isUat: boolean;
};

export const iOSUserAgent =
  "Mozilla/5.0 (iPhone; CPU iPhone OS 14_0_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0 Mobile/15E148 Safari/604.1";

export const defaultUserAgent = Platform.select({
  ios: iOSUserAgent,
  default: undefined
});

export const originSchemasWhiteList = [
  "https://*",
  "iologin://*",
  ...(isDevEnv ? ["http://*"] : [])
];

export const WHITELISTED_DOMAINS = [
  "https://idserver.servizicie.interno.gov.it",
  "https://oidc.idserver.servizicie.interno.gov.it",
  "https://mtls.oidc.idserver.servizicie.interno.gov.it",
  "https://mtls.idserver.servizicie.interno.gov.it",
  "https://ios.idserver.servizicie.interno.gov.it",
  "https://ios.oidc.idserver.servizicie.interno.gov.it"
];

export type CieIdLoginProps = {
  spidLevel: SpidLevel;
  isUat: boolean;
};
