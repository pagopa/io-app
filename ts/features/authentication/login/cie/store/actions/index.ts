export * from "./cie";
export * from "./cieLogin";
