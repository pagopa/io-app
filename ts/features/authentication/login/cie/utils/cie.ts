import cieSdk from "@pagopa/react-native-cie";
import { Platform } from "react-native";

const realIsNfcEnabled = () => cieSdk.isNFCEnabled();
const mockIsNfcEnabled = () => Promise.resolve(true);

const realOpenNfcSettings = cieSdk.openNFCSettings;
const mockOpenNfcSettings = () => Promise.resolve();

const test = Platform.OS !== "android";

export const isNfcEnabled = test ? mockIsNfcEnabled : realIsNfcEnabled;
export const openNFCSettings = test ? mockOpenNfcSettings : realOpenNfcSettings;

export const IO_LOGIN_CIE_SOURCE_APP = "iologincie";
export const IO_LOGIN_CIE_URL_SCHEME = `${IO_LOGIN_CIE_SOURCE_APP}:`;
export const CIE_ID_ERROR = "cieiderror";
export const CIE_ID_ERROR_MESSAGE = "cieid_error_message=";
