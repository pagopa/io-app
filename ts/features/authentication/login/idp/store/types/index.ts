import * as pot from "@pagopa/ts-commons/lib/pot";

export enum ErrorType {
  "LOADING_ERROR" = "LOADING_ERROR",
  "LOGIN_ERROR" = "LOGIN_ERROR"
}

export type RequestInfoPositiveStates = {
  requestState: "LOADING" | "AUTHORIZED" | "AUTHORIZING";
  nativeAttempts: number;
};

export type RequestInfoError = {
  requestState: "ERROR";
  errorType: ErrorType;
  errorCodeOrMessage?: string;
  nativeAttempts: number;
};

export type NativeLoginRequestInfo =
  | RequestInfoPositiveStates
  | RequestInfoError;

export type StandardLoginRequestInfo = {
  requestState: pot.Pot<true, ErrorType>;
};
