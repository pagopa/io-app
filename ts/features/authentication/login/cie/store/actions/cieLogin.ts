import { ActionType, createStandardAction } from "typesafe-actions";
import { SpidLevel } from "../../utils";

export const cieLoginEnableUat = createStandardAction("CIE_LOGIN_ENABLE_UAT")();

export const cieLoginDisableUat = createStandardAction(
  "CIE_LOGIN_DISABLE_UAT"
)();

export const cieIDDisableTourGuide = createStandardAction(
  "CIE_ID_DISABLE_TOUR_GUIDE"
)();
export const cieIDSetSelectedSecurityLevel = createStandardAction(
  "CIE_ID_SET_SELECTED_SECURITY_LEVEL"
)<SpidLevel>();

export type CieLoginConfigActions =
  | ActionType<typeof cieLoginDisableUat>
  | ActionType<typeof cieLoginEnableUat>
  | ActionType<typeof cieIDDisableTourGuide>
  | ActionType<typeof cieIDSetSelectedSecurityLevel>;
