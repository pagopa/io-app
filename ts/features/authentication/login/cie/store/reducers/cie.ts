/**
 * A reducer for the authentication by CIE
 */
import * as pot from "@pagopa/ts-commons/lib/pot";
import { getType } from "typesafe-actions";
import {
  cieIsSupported,
  hasApiLevelSupport,
  hasNFCFeature,
  nfcIsEnabled,
  updateReadingState
} from "../actions";
import { Action } from "../../../../../../store/actions/types";

export type CieState = {
  hasApiLevelSupport: pot.Pot<boolean, Error>;
  hasNFCFeature: pot.Pot<boolean, Error>;
  isCieSupported: pot.Pot<boolean, Error>;
  isNfcEnabled: pot.Pot<boolean, Error>;
  readingEvent: pot.Pot<string, Error>;
};

const INITIAL_STATE: CieState = {
  hasApiLevelSupport: pot.none,
  hasNFCFeature: pot.none,
  isCieSupported: pot.none,
  isNfcEnabled: pot.none,
  readingEvent: pot.none
};

export function cieReducer(
  state: CieState = INITIAL_STATE,
  action: Action
): CieState {
  switch (action.type) {
    case getType(cieIsSupported.success):
      return {
        ...state,
        isCieSupported: pot.some(action.payload)
      };
    case getType(cieIsSupported.failure):
      return {
        ...state,
        isCieSupported: pot.toError(state.isCieSupported, action.payload)
      };
    case getType(hasApiLevelSupport.success):
      return {
        ...state,
        hasApiLevelSupport: pot.some(action.payload)
      };
    case getType(hasApiLevelSupport.failure):
      return {
        ...state,
        hasApiLevelSupport: pot.toError(
          state.hasApiLevelSupport,
          action.payload
        )
      };
    case getType(hasNFCFeature.success):
      return {
        ...state,
        hasNFCFeature: pot.some(action.payload)
      };
    case getType(hasNFCFeature.failure):
      return {
        ...state,
        hasNFCFeature: pot.toError(state.hasNFCFeature, action.payload)
      };
    case getType(nfcIsEnabled.success):
      return {
        ...state,
        isNfcEnabled: pot.some(action.payload)
      };
    case getType(nfcIsEnabled.failure):
      return {
        ...state,
        isNfcEnabled: pot.toError(state.isNfcEnabled, action.payload)
      };

    case getType(updateReadingState.success):
      return {
        ...state,
        readingEvent: pot.some(action.payload)
      };
    case getType(updateReadingState.failure):
      return {
        ...state,
        readingEvent: pot.toError(state.readingEvent, action.payload)
      };

    default:
      return state;
  }
}
