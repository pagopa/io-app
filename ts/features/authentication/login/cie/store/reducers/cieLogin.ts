import { getType } from "typesafe-actions";
import AsyncStorage from "@react-native-async-storage/async-storage";
import {
  createMigrate,
  MigrationManifest,
  PersistConfig,
  PersistedState,
  persistReducer
} from "redux-persist";
import { omit } from "lodash";
import {
  cieIDDisableTourGuide,
  cieIDSetSelectedSecurityLevel,
  cieLoginDisableUat,
  cieLoginEnableUat
} from "../actions";
import { Action } from "../../../../../../store/actions/types";
import { isDevEnv } from "../../../../../../utils/environment";
import { SpidLevel } from "../../utils";

export type CieLoginState = {
  useUat: boolean;
  isCieIDTourGuideEnabled: boolean;
  cieIDSelectedSecurityLevel?: SpidLevel;
};

export const cieLoginInitialState = {
  useUat: false,
  isCieIDTourGuideEnabled: true
};

const cieLoginReducer = (
  state: CieLoginState = cieLoginInitialState,
  action: Action
): CieLoginState => {
  switch (action.type) {
    case getType(cieLoginEnableUat):
      return {
        ...state,
        useUat: true
      };
    case getType(cieLoginDisableUat):
      return {
        ...state,
        useUat: false
      };
    case getType(cieIDDisableTourGuide):
      return {
        ...state,
        isCieIDTourGuideEnabled: false
      };
    case getType(cieIDSetSelectedSecurityLevel):
      return {
        ...state,
        cieIDSelectedSecurityLevel: action.payload
      };
    default:
      return state;
  }
};

const CURRENT_REDUX_CIE_LOGIN_STORE_VERSION = 1;

const migrations: MigrationManifest = {
  "0": (state: PersistedState) => ({
    ...state,
    isCieIDTourGuideEnabled: true
  }),
  /**
   * @param state The slice state
   * Removes `isCieIDFeatureEnabled` from the persist rehydration actions
   */
  "1": (state: PersistedState) => omit(state, "isCieIDFeatureEnabled")
};

const persistConfig: PersistConfig = {
  key: "cieLogin",
  storage: AsyncStorage,
  migrate: createMigrate(migrations, { debug: isDevEnv }),
  version: CURRENT_REDUX_CIE_LOGIN_STORE_VERSION,
  whitelist: ["isCieIDTourGuideEnabled"]
};

export const cieLoginPersistor = persistReducer<CieLoginState, Action>(
  persistConfig,
  cieLoginReducer
);

export const testableCieLoginReducer = isDevEnv
  ? {
      cieLoginReducer,
      migrations
    }
  : undefined;
