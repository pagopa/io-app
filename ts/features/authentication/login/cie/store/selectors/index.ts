import { GlobalState } from "../../../../../../store/reducers/types";

export const isCieLoginUatEnabledSelector = (state: GlobalState) =>
  state.features.loginFeatures.cieLogin.useUat;

export const isCieIDTourGuideEnabledSelector = (state: GlobalState) =>
  state.features.loginFeatures.cieLogin.isCieIDTourGuideEnabled;

export const cieIDSelectedSecurityLevelSelector = (state: GlobalState) =>
  state.features.loginFeatures.cieLogin.cieIDSelectedSecurityLevel;

export const hasNFCFeatureSelector = (state: GlobalState) =>
  state.cie.hasNFCFeature;

export const hasApiLevelSupportSelector = (state: GlobalState) =>
  state.cie.hasApiLevelSupport;

export const isCieSupportedSelector = (state: GlobalState) =>
  state.cie.isCieSupported;

export const isNfcEnabledSelector = (state: GlobalState) =>
  state.cie.isNfcEnabled;

export const readingEventSelector = (state: GlobalState) =>
  state.cie.readingEvent;
