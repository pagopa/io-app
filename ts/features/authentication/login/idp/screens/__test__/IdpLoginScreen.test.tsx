import { createStore } from "redux";
import * as pot from "@pagopa/ts-commons/lib/pot";
import { applicationChangeState } from "../../../../../../store/actions/application";
import { appReducer } from "../../../../../../store/reducers";
import { renderScreenWithNavigationStoreContext } from "../../../../../../utils/testWrapper";
import IdpLoginScreen from "../IdpLoginScreen";
import { AUTHENTICATION_ROUTES } from "../../../../common/navigation/routes";
import * as requestinfo from "../../store/selectors";
import * as IOHooks from "../../../../../../store/hooks";
import * as commonStoreSelector from "../../../../common/store/selectors";
import { SessionToken } from "../../../../../../types/SessionToken";

jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockNavigate
    })
  };
});

const mockNavigate = jest.fn();

jest.mock("../../../../../../hooks/useHeaderSecondLevel", () => ({
  useHeaderSecondLevel: jest.fn()
}));

describe("IdpLoginScreen", () => {
  const mockDispatch = jest.fn();

  jest.spyOn(IOHooks, "useIODispatch").mockReturnValue(mockDispatch);

  jest
    .spyOn(commonStoreSelector, "loggedOutWithIdpAuthSelector")
    .mockReturnValue({
      kind: "LoggedOutWithIdp",
      idp: {
        id: "testidp1",
        name: "testidp1",
        logo: { light: { uri: "" } },
        profileUrl: ""
      },
      reason: "NOT_LOGGED_IN"
    });

  it("should match snapshots", () => {
    const { toJSON } = renderComponent();
    expect(toJSON).toMatchSnapshot();
  });

  it("should render correctly the webview", () => {
    jest
      .spyOn(requestinfo, "standardLoginRequestInfoSelector")
      .mockReturnValue({
        requestState: pot.some(true)
      });

    const { getByTestId } = renderComponent();

    expect(getByTestId("webview-idp-login-screen")).toBeTruthy();
  });

  it("should render correctly the loader", () => {
    jest
      .spyOn(requestinfo, "standardLoginRequestInfoSelector")
      .mockReturnValue({
        requestState: pot.toLoading(pot.none)
      });

    const { getByTestId } = renderComponent();

    expect(getByTestId("loading-indicator")).toBeTruthy();
  });

  it("should render success screen if logged in", () => {
    jest.spyOn(commonStoreSelector, "loggedInAuthSelector").mockReturnValue({
      kind: "LoggedInWithoutSessionInfo",
      idp: {
        id: "testidp1",
        name: "testidp1",
        logo: { light: { uri: "" } },
        profileUrl: ""
      },
      sessionToken: "fake-session-token" as SessionToken,
      _persist: {
        version: -1,
        rehydrated: true
      }
    });

    jest
      .spyOn(requestinfo, "standardLoginRequestInfoSelector")
      .mockReturnValue({
        requestState: pot.some(true)
      });

    const { getByTestId } = renderComponent();

    expect(getByTestId("idp-successful-authentication")).toBeTruthy();
  });
});

const renderComponent = () => {
  const initialState = appReducer(undefined, applicationChangeState("active"));
  const store = createStore(appReducer, initialState as any);

  return renderScreenWithNavigationStoreContext(
    IdpLoginScreen,
    AUTHENTICATION_ROUTES.IDP_LOGIN,
    {},
    store
  );
};
