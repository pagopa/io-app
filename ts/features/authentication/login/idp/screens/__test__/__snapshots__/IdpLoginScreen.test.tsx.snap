// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdpLoginScreen should match snapshots 1`] = `[Function]`;
